pkg/scheduler/algorithm/predicates/utils.go

```diff
@@ -147,3 +148,193 @@ func isCSIMigrationOn(csiNode *storagev1beta1.CSINode, pluginName string) bool {
+	expression := v1.NodeSelectorRequirement{
+		Key:      key,
+		Operator: v1.NodeSelectorOpIn,
+		Values:   vals,
+	}
+	selectorTerm := v1.NodeSelectorTerm{}
+	selectorTerm.MatchExpressions = append(selectorTerm.MatchExpressions, expression)
+	s.NodeSelectorTerms = append(s.NodeSelectorTerms, selectorTerm)
+	return s
+}
+
+func (s *nodeSelectorWrapper) notIn(key string, vals []string) *nodeSelectorWrapper {
+	expression := v1.NodeSelectorRequirement{
+		Key:      key,
+		Operator: v1.NodeSelectorOpNotIn,
+		Values:   vals,
+	}
+	selectorTerm := v1.NodeSelectorTerm{}
+	selectorTerm.MatchExpressions = append(selectorTerm.MatchExpressions, expression)
+	s.NodeSelectorTerms = append(s.NodeSelectorTerms, selectorTerm)
+	return s
+}
+
+func (s *nodeSelectorWrapper) obj() *v1.NodeSelector {
+	return &s.NodeSelector
+}
+
+type labelSelectorWrapper struct{ metav1.LabelSelector }
+
+func makeLabelSelector() *labelSelectorWrapper {
+	return &labelSelectorWrapper{metav1.LabelSelector{}}
+}
+
+func (s *labelSelectorWrapper) label(k, v string) *labelSelectorWrapper {
+	if s.MatchLabels == nil {
+		s.MatchLabels = make(map[string]string)
+	}
+	s.MatchLabels[k] = v
+	return s
+}
+
+func (s *labelSelectorWrapper) in(key string, vals []string) *labelSelectorWrapper {
+	expression := metav1.LabelSelectorRequirement{
+		Key:      key,
+		Operator: metav1.LabelSelectorOpIn,
+		Values:   vals,
+	}
+	s.MatchExpressions = append(s.MatchExpressions, expression)
+	return s
+}
+
+func (s *labelSelectorWrapper) notIn(key string, vals []string) *labelSelectorWrapper {
+	expression := metav1.LabelSelectorRequirement{
+		Key:      key,
+		Operator: metav1.LabelSelectorOpNotIn,
+		Values:   vals,
+	}
+	s.MatchExpressions = append(s.MatchExpressions, expression)
+	return s
+}
+
+func (s *labelSelectorWrapper) exists(k string) *labelSelectorWrapper {
+	expression := metav1.LabelSelectorRequirement{
+		Key:      k,
+		Operator: metav1.LabelSelectorOpExists,
+	}
+	s.MatchExpressions = append(s.MatchExpressions, expression)
+	return s
+}
+
+func (s *labelSelectorWrapper) notExist(k string) *labelSelectorWrapper {
+	expression := metav1.LabelSelectorRequirement{
+		Key:      k,
+		Operator: metav1.LabelSelectorOpDoesNotExist,
+	}
+	s.MatchExpressions = append(s.MatchExpressions, expression)
+	return s
+}
+
+func (s *labelSelectorWrapper) obj() *metav1.LabelSelector {
+	return &s.LabelSelector
+}
+
+type podWrapper struct{ v1.Pod }
+
+func makePod() *podWrapper {
+	return &podWrapper{v1.Pod{}}
+}
+
+func (p *podWrapper) obj() *v1.Pod {
+	return &p.Pod
+}
+
+func (p *podWrapper) name(s string) *podWrapper {
+	p.Name = s
+	return p
+}
+
+func (p *podWrapper) namespace(s string) *podWrapper {
+	p.Namespace = s
+	return p
+}
+
+func (p *podWrapper) node(s string) *podWrapper {
+	p.Spec.NodeName = s
+	return p
+}
+
+func (p *podWrapper) nodeSelector(m map[string]string) *podWrapper {
+	p.Spec.NodeSelector = m
+	return p
+}
+
+// represents HARD node affinity in particular
```

Comment:
comment is cut short, also please end all your comments with a period.

Reply:
Done.